### PR TITLE
Prototype for CBA

### PR DESCRIPTION
**Pull request by coreyhulen**
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10127

### CODE REVIEW

**Comment by coreyhulen** [comment]
:+1: I cannot officially review it because it's my branch, but your changes look good.

File: api4/user.go [hunk 1]
```diff
@@ -942,8 +942,29 @@ func login(c *Context, w http.ResponseWriter, r *http.Request) {
 	deviceId := props["device_id"]
 	ldapOnly := props["ldap_only"] == "true"
 
+	certPem, certSubject, certEmail := c.App.CheckForClienSideCert(r)
+	mlog.Info(fmt.Sprintf("Client Cert PEM: %v", certPem))
```
Comment (crspeller) [inline api4/user.go]: We probably want this on debug. Also avoid the Sprintf by using a field. (you can combine these three lines into one)

File: api4/user.go [hunk 2]
```diff
@@ -942,8 +942,29 @@ func login(c *Context, w http.ResponseWriter, r *http.Request) {
 	deviceId := props["device_id"]
 	ldapOnly := props["ldap_only"] == "true"
 
+	certPem, certSubject, certEmail := c.App.CheckForClienSideCert(r)
```
Comment (crspeller) [inline api4/user.go]: This should be inside the experimental settings block.

File: app/login.go [hunk 3]
```diff
@@ -35,6 +53,13 @@ func (a *App) AuthenticateUserForLogin(id, loginId, password, mfaToken string, l
 		return nil, err
 	}
 
+	// If client side cert is enable and it's checking as a primary source
+	// then trust the proxy and cert that the correct user is supplied and allow
+	// them access
+	if *a.Config().ExperimentalSettings.ClientSideCertEnable && *a.Config().ExperimentalSettings.ClientSideCertCheck == model.CLIENT_SIDE_CERT_CHECK_PRIMARY_AUTH {
```
Comment (crspeller) [inline app/login.go]: Can I get some materials to read on how this is supposed to work. We don't appear to be validating this at all. How do I know the email matches the cert provided? Can't I just login as anyone?

Comment (crspeller) [reply app/login.go]: Resolved with explanation here: https://pre-release.mattermost.com/core/pl/fk9omytpopgm5qt5r5s1r14cfc

**Comment by enahum** [comment]
@coreyhulen this is your PR, do you want to make the requested changes?

**Comment by coreyhulen** [comment]
Yeah, I'll make the changes, plus I should add some unit tests :)

**Comment by coreyhulen** [comment]
OK @crspeller and @enahum please re-review.  I'ved fixed the issues mentioned and added some unit tests.

File: api4/user.go [hunk 4]
```diff
@@ -983,8 +983,28 @@ func login(c *Context, w http.ResponseWriter, r *http.Request) {
 	deviceId := props["device_id"]
 	ldapOnly := props["ldap_only"] == "true"
 
+	if *c.App.Config().ExperimentalSettings.ClientSideCertEnable {
+		if license := c.App.License(); license == nil || !*license.Features.SAML {
+			c.Err = model.NewAppError("ClientSideCertNotAllowed", "Attempt to use the experimental feature ClientSideCertEnable without a valid enterprise license", nil, "", http.StatusBadRequest)
+			return
+		} else {
+			certPem, certSubject, certEmail := c.App.CheckForClienSideCert(r)
+			mlog.Debug(fmt.Sprintf("Client Cert Subject: %v", certSubject))
```
Comment (crspeller) [inline api4/user.go]: I think we would prefer 
```go
mlog.Debug("Client Cert", mlog.String("cert_subject", certSubject), mlog.String("cert_email", certEmail))
```

File: api4/user_test.go [hunk 5]
```diff
@@ -2235,6 +2235,58 @@ func TestSetProfileImage(t *testing.T) {
...
+
+	user, resp := Client.Login(th.BasicUser.Email, th.BasicUser.Password)
+	if resp.Error.StatusCode != 400 && user == nil {
+		t.Fatal("Should have failed because it's missing the cert header")
+	}
+
+	Client.HttpHeader["X-SSL-Client-Cert"] = "valid_cert_fake"
+	user, resp = Client.Login(th.BasicUser.Email, th.BasicUser.Password)
+	if resp.Error.StatusCode != 400 && user == nil {
+		t.Fatal("Should have failed because it's missing the cert header")
```
Comment (crspeller) [inline api4/user_test.go]: Because the cert is invalid.

File: api4/user_test.go [hunk 6]
```diff
@@ -2235,6 +2235,58 @@ func TestSetProfileImage(t *testing.T) {
...
+	}
+
+	Client.HttpHeader["X-SSL-Client-Cert-Subject-DN"] = "C=US, ST=Maryland, L=Pasadena, O=Brent Baccala, OU=FreeSoft, CN=www.freesoft.org/emailAddress=mis_match" + th.BasicUser.Email
+	user, resp = Client.Login(th.BasicUser.Email, "")
+	if resp.Error.StatusCode != 400 && user == nil {
+		t.Fatal("Should have failed because it's missing the cert header")
```
Comment (crspeller) [inline api4/user_test.go]: Because the email is mismatched.

File: api4/user_test.go [hunk 7]
```diff
@@ -2235,6 +2235,58 @@ func TestSetProfileImage(t *testing.T) {
...
+
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.ExperimentalSettings.ClientSideCertEnable = true
+		*cfg.ExperimentalSettings.ClientSideCertCheck = model.CLIENT_SIDE_CERT_CHECK_SECONDARY_AUTH
+	})
+
+	Client.HttpHeader["X-SSL-Client-Cert-Subject-DN"] = "C=US, ST=Maryland, L=Pasadena, O=Brent Baccala, OU=FreeSoft, CN=www.freesoft.org/emailAddress=" + th.BasicUser.Email
+	user, resp = Client.Login(th.BasicUser.Email, "")
+	if resp.Error.StatusCode != 400 && user == nil {
+		t.Fatal("Should have failed because it's missing the cert header")
```
Comment (crspeller) [inline api4/user_test.go]: Because the secondary auth password is missing.

**Comment by coreyhulen** [comment]
ok @crspeller ready for re-review

**Pull request merged**